### PR TITLE
State Trie Migration : Introduce state migration progress

### PR DESCRIPTION
**Pull request by KimKyungup**
## Proposed changes

This PR introduces state migration progress feature.
This feature prints the log like below.
```
INFO[05/15,16:17:30 +09] [47] Trie sync progress by depth #0            committed=0    pending=1    progress=0.000
INFO[05/15,16:17:30 +09] [47] Trie sync progress by depth #1            committed=4    pending=16   progress=25.000
INFO[05/15,16:17:30 +09] [47] Trie sync progress by depth #2            committed=240  pending=256  progress=93.750
INFO[05/15,16:17:30 +09] [47] Trie sync progress by depth #3            committed=3782 pending=3798 progress=99.579
INFO[05/15,16:17:30 +09] [47] Trie sync progress by depth #4            committed=8966 pending=8982 progress=99.822
INFO[05/15,16:17:30 +09] [47] Trie sync progress by depth #5            committed=1479 pending=1485 progress=99.596
INFO[05/15,16:17:30 +09] [47] Trie sync progress by depth #6            committed=121  pending=122  progress=99.180
INFO[05/15,16:17:30 +09] [47] Trie sync progress by depth #7            committed=8    pending=8    progress=100.000
INFO[05/15,16:17:30 +09] [47] Trie sync progress                        progress=99.82186595413049%
WARN[05/15,16:17:30 +09] [5] State migration progress                  progress=99.8219%            committedCnt=97869 pendingCnt=89994 read=76783 readElapsed=538.888652ms written=61652 writeElapsed=99.547522ms elapsed=1.129802835s
```

And the progress information can be checked in API like below.
```
> admin.statusStateMigration
{
  committed: 36932,
  isMigration: true,
  migrationBlockNumber: 2436096,
  pending: 99631,
  progress: 99.31791537887669
}
```

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- Please leave the issue numbers or links related to this PR here.

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: storage/statedb/sync.go [hunk 1]
```diff
@@ -338,3 +350,53 @@ func (s *TrieSync) commit(req *request) (err error) {
...
+	for i := 0; i < 20; i++ {
+		var progressByDepth float64
+		c, p := s.CommittedByDepth(i), s.PendingByDepth(i)
+		if p == 0 {
+			break
+		}
+		if p > 0 {
+			progressByDepth = float64(c) / float64(p) * 100
+		}
+		logger.Info("Trie sync progress by depth #"+strconv.Itoa(i), "committed", c, "pending", p, "progress", progressByDepth)
```
Comment (KimKyungup) [inline storage/statedb/sync.go:395]: This log level will be changed to "debug" in future PR after developing State Migration.

File: storage/statedb/sync.go [hunk 2]
```diff
@@ -338,3 +350,53 @@ func (s *TrieSync) commit(req *request) (err error) {
...
+	}
+
+	progress *= 100
+	logger.Info("Trie sync progress ", "progress", strconv.FormatFloat(progress, 'f', -1, 64)+"%")
```
Comment (KimKyungup) [inline storage/statedb/sync.go:399]: This log level will be changed to "debug" in future PR after developing State Migration.

**Comment by KimKyungup** [comment]
@aidan-kwon @winnie-byun I rebased. PTAL.

File: storage/statedb/sync.go [hunk 3]
```diff
@@ -338,3 +350,53 @@ func (s *TrieSync) commit(req *request) (err error) {
...
+
+	for i := 0; i < 6; i++ {
+		c, p := s.CommittedByDepth(i), s.PendingByDepth(i)
+		if p == 0 {
+			break
+		}
+		newProgress := float64(c) / float64(p)
+		if newProgress > progress {
+			progress = newProgress
+		}
```
Comment (aidan-kwon) [inline storage/statedb/sync.go]: I think I can be moved to the next `for` loop.

File: storage/statedb/sync.go [hunk 4]
```diff
@@ -82,19 +83,23 @@ type StateTrieReadDB interface {
...
-	database StateTrieReadDB          // Persistent database to check for existing entries
-	membatch *syncMemBatch            // Memory buffer to avoid frequest database writes
-	requests map[common.Hash]*request // Pending requests pertaining to a key hash
-	queue    *prque.Prque             // Priority queue with the pending requests
+	database         StateTrieReadDB          // Persistent database to check for existing entries
+	membatch         *syncMemBatch            // Memory buffer to avoid frequest database writes
+	requests         map[common.Hash]*request // Pending requests pertaining to a key hash
+	queue            *prque.Prque             // Priority queue with the pending requests
+	retrievedByDepth map[int]int              // Retrieved trie count by depth
+	committedByDepth map[int]int              // Committed trie count by depth
```
Comment (aidan-kwon) [inline storage/statedb/sync.go:91]:
```suggestion
	retrievedByDepth map[int]int              // Retrieved trie node number counted by depth
	committedByDepth map[int]int              // Committed trie nodes number counted by depth
```

File: storage/statedb/sync.go (hunk 3 above)
Comment (KimKyungup) [reply storage/statedb/sync.go]: Good point. :)

File: storage/statedb/sync.go [hunk 5]
```diff
@@ -338,3 +350,44 @@ func (s *TrieSync) commit(req *request) (err error) {
 	}
 	return nil
 }
+
+// RetrievedByDepth returns the retrieved trie count by given depth.
```
Comment (winnie-byun) [inline storage/statedb/sync.go:354]:
```suggestion
// RetrievedByDepth returns the retrieved trie count by given depth.
// This number is same as the number of nodes that needs to be commited to complete trie sync. 
```

File: blockchain/blockchain.go [hunk 6]
```diff
@@ -447,7 +461,7 @@ func (bc *BlockChain) restartStateMigration() {
 		}
 
 		root := block.Root()
-		logger.Warn("Restart state trie migration", "blockNumber", number, "root", root.String())
+		logger.Warn("State migration is restarted", "blockNumber", number, "root", root.String())
```
Comment (winnie-byun) [inline blockchain/blockchain.go:464]:
```suggestion
		logger.Warn("Continue to migration state trie", "blockNumber", number, "root", root.String())
```

Comment (KimKyungup) [reply blockchain/blockchain.go:464]: How about this?
I intended use "State migration` for consisenty of relevant logs.
```suggestion
		logger.Warn("State migration is continued", "blockNumber", number, "root", root.String())
```

Comment (winnie-byun) [reply blockchain/blockchain.go:464]: I think `restart` and `continue` is similar. We can go both ways. :)

Comment (KimKyungup) [reply blockchain/blockchain.go:464]: Ok. Off we go~!

**Pull request merged**